### PR TITLE
[doc] add doc and example for Ray TLS

### LINKED ISSUE #889: [Bug] RAY_USE_TLS in Kuberay
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

Others

### What happened + What you expected to happen

Hi. Regarding **RAY_USE_TLS** in Ray on k8s. I've defined env vars in the helm chart values of the head node.

**Ray image v2.2.0**

All of the files are mounted and existing in `/tmp/ray/certs/` as expected. During the head node launch I have this error message:

```
details = "failed to connect to all addresses; last error: UNKNOWN: ipv4:10.108.163.118:6379: Peer name 10.108.163.118 is not in peer certificate"
	debug_error_string = "UNKNOWN:failed to connect to all addresses; last error: UNKNOWN: ipv4:10.108.163.118:6379: Peer name 10.108.163.118 is not in peer certificate {grpc_status:14, created_time:"2023-01-31T00:35:58.84152496-08:00"}"
```
**where 10.108.163.118 - pod ip.**

I added in SAN: 
`IP.1 = 127.0.0.1`
`DNS.1 = localhost `

But seems like it tries to connect using pod IP. But I cannot define in SAN pod IP cause its dynamic and will be changed during the restart. 
Any suggestions/workarounds? 
Thanks.

### Reproduction script

TLS env vars in the helm chart:
    - name: RAY_USE_TLS
       value: "1"
    - name: RAY_TLS_SERVER_CERT
       value: "/tmp/ray/certs/server.crt"
    - name: RAY_TLS_SERVER_KEY
       value: "/tmp/ray/certs/server.key"
    - name: RAY_TLS_CA_CERT
       value: "/tmp/ray/certs/rootCA.crt"

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by YQ-Wang**
## Why are these changes needed?
This PR gives a detailed doc and a step-by-step guide on how to configure the TLS Auth for the Ray cluster.

## Related issue number
Closes https://github.com/ray-project/kuberay/issues/889

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Test locally in a kind cluster to validate the TLS logic
   - [ ] Unit tests
   - [ ] Release tests
   - [] This PR is not tested :(

### CODE REVIEW

**Comment by YQ-Wang** [comment]
@kevin85421

File: doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml
Comment (kevin85421) [inline doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml:225]: `IfNotPresent`?

**Review by kevin85421 — COMMENTED** [review]
Overall looks good to me. Will approve it when I have time to test it manually.

**Review by architkulkarni — COMMENTED** [review]
The doc looks good! I'll leave it to @kevin85421 to run the manual test

File: doc/source/ray-core/configure.rst
Comment (YQ-Wang) [reply doc/source/ray-core/configure.rst]: Done

File: doc/source/ray-core/configure.rst
Comment (kevin85421) [inline doc/source/ray-core/configure.rst:176]: @architkulkarni's comments from https://github.com/ray-project/kuberay/pull/989

https://github.com/ray-project/kuberay/pull/989#discussion_r1149634284
https://github.com/ray-project/kuberay/pull/989#discussion_r1149635676

Comment (YQ-Wang) [reply doc/source/ray-core/configure.rst]: Done

File: doc/source/ray-core/configure.rst
Comment (tedhtchang) [inline doc/source/ray-core/configure.rst:229]:
```suggestion
and the `csr.conf` configuration file. Finally, a self-signed certificate (`tls.crt`) is created using
```

Comment (YQ-Wang) [reply doc/source/ray-core/configure.rst]: Done

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:177]:
```suggestion
various processes (client, head, workers) are encrypted.
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:179]:
```suggestion
In TLS, the private key and public key are used for encryption and decryption.
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:180]:
```suggestion
The former is kept secret by the owner and the latter is shared with the other
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:181]:
```suggestion
party. This pattern ensures that only the intended recipient can read the message.
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:189]:
```suggestion
To obtain a digital certificate, the owner of the public key must generate a
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:194]:
```suggestion
Here is a step-by-step guide for adding TLS Authentication to a static Kubernetes Ray cluster using a
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:218]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:222]:
```suggestion
includes two shell scripts: `gencert_head.sh` and `gencert_worker.sh`. These scripts
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:221]:
```suggestion
<https://raw.githubusercontent.com/ray-project/ray/master/doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml>`__, has a ConfigMap named `tls` that
```

**Comment by YQ-Wang** [comment]
@angelinalg done for all, thanks for reviewing!

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:237]:
```suggestion
- ``RAY_TLS_SERVER_CERT``: Location of a `certificate file (tls.crt)`, which is presented to other endpoints to achieve mutual authentication.
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:238]:
```suggestion
- ``RAY_TLS_SERVER_KEY``: Location of a `private key file (tls.key)`, which is the cryptographic means to prove to other endpoints that you are the authorized user of a given certificate.
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:239]:
```suggestion
- ``RAY_TLS_CA_CERT``: Location of a `CA certificate file (ca.crt)`, which allows TLS to decide whether an endpoint's certificate has been signed by the correct authority.
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:250]:
```suggestion
  # is established successfully
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:253]:
```suggestion
  # Since service-ray-head hasn't added to the alt_names section in the certificate, the connection fails and an error
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:254]:
```suggestion
  # message similar to the following is displayed: "Peer name service-ray-head is not in peer certificate".
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:264]:
```suggestion
The exact overhead depends on the nature of your workload.
```

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:257]:
```suggestion
  # After you add `DNS.3 = service-ray-head` to the alt_names sections and deploy the YAML again, the connection is able to work.
```

**Comment by angelinalg** [comment]
Added some copy edits. Thanks for updating the docs!

**Comment by YQ-Wang** [comment]
> Added some copy edits. Thanks for updating the docs!

@angelinalg Done for all, thanks!

**Review by kevin85421 — APPROVED** [review]
I tested this PR on my devbox. 

```sh
kind create cluster --image=kindest/node:v1.23.0
kubectl apply -f doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml
kubectl get pods # 3 running Pods (1 head Pod, 2 worker Pods)
```

File: doc/source/ray-core/configure.rst
Comment (tedhtchang) [inline doc/source/ray-core/configure.rst:216]:
```suggestion
# Alternatively, the command automatically encode and create the secret for the CA keypair.
kubectl create secret generic ca-tls --from-file=ca.crt=<path-to-ca.crt> --from-file=ca.key=<path-to-ca.key>

```

File: doc/source/ray-core/configure.rst
Comment (tedhtchang) [inline doc/source/ray-core/configure.rst:231]: no biggie. no one will notice. change up to you.

```suggestion
the Certificate Authority's (`ca.key and ca.crt`) keypair and the CSR (`ca.csr`).
```

**Review by architkulkarni — APPROVED** [review]

**Comment by architkulkarni** [comment]
@angelinalg if the updates look okay, could you please approve this on behalf of ray-docs?

File: doc/source/ray-core/configure.rst
Comment (angelinalg) [inline doc/source/ray-core/configure.rst:224]:
```suggestion
includes two shell scripts: `gencert_head.sh` and `gencert_worker.sh`. These scripts produce the private key
```

File: doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml
Comment (angelinalg) [inline doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml:180]:
```suggestion
      # nodes) goes down as well. If you want Kubernetes to bring up a new
```

File: doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml
Comment (angelinalg) [inline doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml:186]:
```suggestion
      # object store. If you do not provide this, Ray falls back to
```

File: doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml
Comment (angelinalg) [inline doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml:235]:
```suggestion
          # object store. If you do not provide this, Ray falls back to
```

File: doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml
Comment (angelinalg) [inline doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml:341]:
```suggestion
        # object store. If you do not provide this, Ray falls back to
```

**Review by angelinalg — APPROVED** [review]
LGTM

**Comment by kevin85421** [comment]
@architkulkarni @angelinalg Can we merge this PR?

**Comment by architkulkarni** [comment]
@YQ-Wang Potentially related error, could you please take a look? https://buildkite.com/ray-project/oss-ci-build-pr/builds/18218#01877278-0e2e-4284-a496-a0f75323aa22/1845-2245
`/ray/doc/source/ray-core/configure.rst:216: WARNING: Explicit markup ends without a blank line; unexpected unindent.`

**Comment by YQ-Wang** [comment]
> ray-core/configure.rst:216

@architkulkarni Done, PTAL

**Comment by architkulkarni** [comment]
Windows failures
- test_task_events unrelated
- test_client_library_integration unrelated
Linkcheck failure "404 Client Error: Not Found for url: https://raw.githubusercontent.com/ray-project/ray/master/doc/source/cluster/kubernetes/configs/static-ray-cluster.tls.yaml"  https://buildkite.com/ray-project/oss-ci-build-pr/builds/20462#0187ddaa-ae57-4c21-924e-89ba9c40561d/870-4116 is related, but should pass once this PR is merged.
Doctest failure `AttributeError: 'RayTrialExecutor' object has no attribute '_used_resources_string'` unrelated
ML test `simple_tune` unrelated

**Pull request merged**
